File: 2022/2/day2.hs
import Data.Char (ord)

toIDs :: [Char] -> [Int]
toIDs [ca, cb] = [(ord ca) - (ord 'A'), (ord cb) - (ord 'X')]

score :: [Int] -> Int
score [a,b] = b + 1 + 3 * (mod (b - a + 1) 3)

score2 :: [Int] -> Int
score2 [a,o] = (mod (a + o - 1) 3) + 1 + 3 * o

main :: IO()
main = do
    content <- readFile "input.txt"
    let inp = map(toIDs . map head . words) . lines $ content
    solve score inp
    solve score2 inp
    where solve sc = print . sum . map sc

File: 2022/2/day2.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ifstream f("input.txt");
    char ca, cb;
    int s1 = 0, s2 = 0;
    while (f >> ca >> cb) {
        int a = ca - 'A', b = cb - 'X';
        s1 += b + 1 + 3 * ((b - a + 4) % 3);
        s2 += (a + b + 2) % 3 + 1 + 3 * b;
    }
    cout << s1 << ' ' << s2;
}